code/datums/browser.dm

```diff
@@ -100,11 +100,114 @@
 		window_size = "size=[width]x[height];"
 	user << browse(get_content(), "window=[window_id];[window_size][window_options]")
 	if (use_onclose)
-		onclose(user, window_id, ref)
+		spawn(0)
+			//winexists sleeps, so we don't need to.
+			for (var/i in 1 to 10)
+				if (user && winexists(user, window_id))
+					onclose(user, window_id, ref)
+					break
+
 
 /datum/browser/proc/close()
 	user << browse(null, "window=[window_id]")
 
+/datum/browser/alert
+	var/selectedbutton = 0
+	var/opentime = 0
+	var/timeout
+	var/stealfocus
+
+/datum/browser/alert/New(User,Message,Title,Button1="Ok",Button2,Button3,StealFocus = 1,Timeout=6000)
+	if (!User)
+		return
+
+	var/output =  {"<center><b>[Message]</b></center><br />
+		<div style="text-align:center">
+		<a style="font-size:large;float:[( Button2 ? "left" : "right" )]" href="?src=\ref[src];button=1">[Button1]</a>"}
+
+	if (Button2)
+		output += {"<a style="font-size:large;[( Button3 ? "" : "float:right" )]" href="?src=\ref[src];button=2">[Button2]</a>"}
+
+	if (Button3)
+		output += {"<a style="font-size:large;float:right" href="?src=\ref[src];button=3">[Button3]</a>"}
+
+	output += {"</div>"}
+
+	..(User, ckey("[User]-[Message]-[Title]-[world.time]-[rand(1,10000)]"), Title, 350, 150, src)
+	set_content(output)
+	stealfocus = StealFocus
+	if (!StealFocus)
+		window_options += "focus=false;"
+	timeout = Timeout
+
+/datum/browser/alert/open()
+	set waitfor = 0
+	opentime = world.time
+
+	if (stealfocus)
+		. = ..(use_onclose = 1)
+	else
+		var/focusedwindow = winget(user, null, "focus")
+		. = ..(use_onclose = 1)

```

Comment:
because focusedwindow needs to be set BEFORE we open the window. other wise we risk that being set to the window we just opened. but winget sleeps and so there is no need to sleep if we don't need the focus'ed window.